presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -290,11 +291,13 @@ else if (conjunct instanceof ComparisonExpression) {
                     if (firstDependencies.stream().allMatch(left::canResolve) && secondDependencies.stream().allMatch(right::canResolve)) {
                         leftComparisonExpressions.add(firstExpression);
                         rightComparisonExpressions.add(secondExpression);
+                        addNullFilters(complexJoinExpressions, node.getType(), firstExpression, secondExpression);
                         joinConditionComparisonOperators.add(comparisonOperator);
                     }
                     else if (firstDependencies.stream().allMatch(right::canResolve) && secondDependencies.stream().allMatch(left::canResolve)) {
                         leftComparisonExpressions.add(secondExpression);
                         rightComparisonExpressions.add(firstExpression);
+                        addNullFilters(complexJoinExpressions, node.getType(), secondExpression, firstExpression);
```

Comment:
Are you sure these are just the keys not any other expressions?

Reply:
@Kaikalur These can be expressions as well, which can be completely resolved from left and right. For example, for the failing test query
```
SELECT DISTINCT a.orderkey FROM 
(SELECT CASE WHEN orderkey > 10 THEN orderkey END orderkey FROM orders WHERE orderkey < 100) a 
RIGHT OUTER JOIN 
(SELECT * FROM orders WHERE orderkey < 100) b ON a.orderkey = b.orderkey
```

I am expecting optimised version as

```
SELECT DISTINCT a.orderkey FROM 
(SELECT CASE WHEN orderkey > 10 THEN orderkey END orderkey FROM orders WHERE orderkey < 100) a 
RIGHT OUTER JOIN 
(SELECT * FROM orders WHERE orderkey < 100) b ON a.orderkey = b.orderkey
WHERE CASE WHEN orderkey > 10 THEN orderkey END IS NOT NULL
```
Which is okay IMO, because `CASE WHEN orderkey > 10 THEN orderkey END IS NOT NULL` can get pushed down to scan operator for `orders` table.

Reply:
This might end up introducing a regression, especially when `NULL`s are rare and/or expression is expensive. I think a better way to implement this is to do this in plan optimization (add an additional filter node after project to avoid recompute the expression). If nothing else, the default of this flag should be `false`. Maybe we should run some benchmarking first.

Reply:
As discussed offline, we were thinking of doing performance testing. In the meantime, planned CSE changes will solve this problem. 

We will roll out this optimisation in disabled state. And when CSE changes are released, we will re-enable this.